### ISSUE TITLE
VS Code doesn't read Git packed-refs

### ISSUE THREAD

**Issue by mankewd** [microsoft/vscode#79776]
VS Code doesn't seem to read the Git packed-refs file for remote branch information. It only looks in the .git/refs/remotes/origin folder but if the Git repository has a lot of seldom used branches then the hash for that branch will be in the packed-refs file and not in the above folder. Consequently, checking out one of those branches will lead to continuous file watcher errors. For example:

```
Failed to watch ref 'd:\apps\wwwroot\www\.git\refs\remotes\origin\[remote_branch_name]', is most likely packed.
Error: ENOENT: no such file or directory, watch 'd:\apps\wwwroot\www\.git\refs\remotes\origin\john_ifims_remove_eforms'
	at FSWatcher.start (internal/fs/watchers.js:164:26)
	at Object.watch (fs.js:1232:11)
	at Object.t.watch (c:\Program Files\Microsoft VS Code\resources\app\extensions\git\dist\main.js:1:6950)
	at E.updateTransientWatchers (c:\Program Files\Microsoft VS Code\resources\app\extensions\git\dist\main.js:1:97463)
	at u.fire (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:50:227)
	at T.updateModelState (c:\Program Files\Microsoft VS Code\resources\app\extensions\git\dist\main.js:1:114843)
```

VS Code has correctly guessed that the reference is packed but I'm not aware of any way to force Git to unpack references and I may not want to update the branch I've checked out and commit the change in order for Git to create the individual reference file into the refs/remotes/origin folder. This is either a bug or a feature request but I think VS Code should really have the ability to look for Git references in the packed-refs file. 

Thanks

Version: 1.37.1 (system setup)
Commit: f06011ac164ae4dc8e753a3fe7f9549844d15e35
Date: 2019-08-15T16:17:55.855Z
Electron: 4.2.7
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.17134

**Comment by joaomoreno** [comment]
When you say _ doesn't read Git packed-refs_, where in the product does that surface? What feature is broken, exactly?

**Comment by mankewd** [comment]
Hi Jaoa,

Thanks for looking into this. When the Git remote branch reference is stored by Git only in the packed-refs file, VS Code can't sync the local branch with its associated upstream branch. This seems to be due to VS Code only looking for Git references in the refs/remotes/origin folder and not in the packed-refs file in the .git directory. Consequently, every time VS Code attempts to sync (about every minute it seems)  an exception is thrown into the Git log stating that it can't find the reference and that it may be packed. That much is true so it would be an improvement in VS Code's Git integration if it could read the packed-refs file instead of relying solely on the individual branch references being available in the refs/remotes/origin folder.

Regards,

Don

[https://ipmcdn.avast.com/images/icons/icon-envelope-tick-green-avg-v1.png]<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>     Virus-free. www.avg.com<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>

**Comment by joaomoreno** [comment]
VS Code spawns git for everything git-related. It doesn't look into any folders manually.

> Consequently, every time VS Code attempts to sync (about every minute it seems)  an exception is thrown into the Git log stating that it can't find the reference and that it may be packed.

Can you show me the exception?

Can you record the issue in a screen recording?

**Comment by mankewd** [comment]
The exception should be included in the information I submitted on the issue I raised.

It's late here in Australia so I've shut down my coding environment for the day. I can reproduce the issue and send you more of the Git log in a day or so. I'm not sure a screen recording would be much help.because everything should be in the log.

________________________________

**Comment by joaomoreno** [comment]
Got it. That error is only for file watching. It should not impact the features at all. Is there a feature which does not work?

**Comment by mankewd** [comment]
That's a good question. I was under the impression that the exception was relevant to Git's ability to sync the local branch to the remote (i.e. to able to push, pull or publish). If nothing else having the exception thrown to the log every minute is distracting.

But I was able to commit and push a small change to the branch I had checked out to the remote branch. Then Git created a reference in refs/remotes/origin/ and the exception went away. However, there are times where I may want to check out a remote branch and not make any changes and that would lead to those continuous file watcher exceptions if there is no individual reference file for it.

I haven't been able to establish conclusively if that exception breaks anything but I thought I would raise it as issue because exceptions are generally indicative of a problem and I'm by no means a Git or VS Code expert. When I saw the exception I thought there was a problem and I spent some time trying to track it down. If having raised an issue saves someone else some time then it was worthwhile even if nothing is actually broken as it's a bit of a red herring.

As I think I mentioned in the issue, I'm not aware of any way to unpack the packed-refs file and if VS Code spawns Git for all of its interactions, why would Git not look in the packed-refs file?  Is that actually a Git issue and the exception that VS Code is dumping into the log is actually a Git exception that has bubbled back up to VS Code so it displays it?

________________________________
